File: frontend/src/api/feedback.ts
import { apiFetch } from './client'

export async function submitRating(data: {
  pdf_title: string
  rating: number
  comment?: string
  result_json?: Record<string, unknown>
  search_set_uuid?: string
}): Promise<{ complete: boolean }> {
  return apiFetch('/api/feedback/submit_rating', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export async function submitChatFeedback(data: {
  conversation_uuid?: string
  message_index?: number
  rating: 'up' | 'down'
  comment?: string
}): Promise<{ complete: boolean }> {
  return apiFetch('/api/feedback/chat', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

// Free-form positive / idea feedback that is NOT a support ticket — it never
// enters the triage queue. Backs the support-panel "something that's working"
// affordance.
export async function submitProductFeedback(data: {
  message: string
  sentiment?: 'positive' | 'idea'
  source?: string
  feature?: string
}): Promise<{ complete: boolean }> {
  return apiFetch('/api/feedback/product', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export type PositiveFeedbackItem = {
  source: 'chat' | 'extraction' | 'product'
  sentiment: string
  message: string | null
  feature: string | null
  user_id: string | null
  created_at: string | null
}

export type PositiveFeedbackStats = {
  by_source: { chat: number; extraction: number; product: number }
  thumbs_up_rate: number | null
  positive_last_7_days: number
}

export function listPositiveFeedback(
  source?: 'chat' | 'extraction' | 'product',
  limit = 50,
): Promise<{ items: PositiveFeedbackItem[]; count: number }> {
  const params = new URLSearchParams()
  params.set('limit', String(limit))
  if (source) params.set('source', source)
  return apiFetch(`/api/feedback/admin/positive?${params}`)
}

export function getPositiveFeedbackStats(): Promise<PositiveFeedbackStats> {
  return apiFetch('/api/feedback/admin/stats')
}

File: backend/app/models/feedback.py
"""Extraction quality feedback model."""

import datetime
from typing import Optional

from beanie import Document
from pydantic import Field


class ChatFeedback(Document):
    conversation_uuid: Optional[str] = None
    message_index: Optional[int] = None
    rating: str = "up"  # "up" or "down"
    comment: Optional[str] = None
    user_id: Optional[str] = None
    # KB the rated message answered from (when known — supplied by the chat
    # surface when the message used RAG). Phase 5 of the loop-closure plan:
    # an elevated thumbs-down rate per KB auto-enqueues a shadow KB
    # autovalidate run so the feedback signal becomes input to the optimizer
    # instead of dying in this collection.
    kb_uuid: Optional[str] = None
    created_at: datetime.datetime = Field(default_factory=datetime.datetime.utcnow)

    class Settings:
        name = "chat_feedback"
        indexes = [
            "conversation_uuid",
            "kb_uuid",
            [("kb_uuid", 1), ("created_at", -1)],
        ]


class ExtractionQualityRecord(Document):
    pdf_title: str
    star_rating: int = Field(ge=1, le=5)
    comment: Optional[str] = None
    result_json: Optional[str] = None
    user_id: Optional[str] = None
    search_set_uuid: Optional[str] = None
    created_at: datetime.datetime = Field(default_factory=datetime.datetime.utcnow)

    class Settings:
        name = "extraction_quality_record"


class ProductFeedback(Document):
    """Free-form, non-negative product feedback that isn't tied to a single
    chat message or extraction — a user telling us what's working, or an idea.

    Negatives deliberately stay on the support-ticket rails (they need triage,
    assignment, and a status that can close). This collection is the home for
    positive/idea signal so it stops dying write-only: the "What's Working"
    admin surface reads it alongside thumbs-up chat feedback and high extraction
    star ratings.
    """

    # "positive" = something worked well; "idea" = a suggestion offered warmly,
    # not a defect report. No "negative" — that path creates a support ticket.
    sentiment: str = "positive"
    message: str
    # Where it was captured, so the admin feed can group by surface and we can
    # add new capture points without schema churn (e.g. "support_panel").
    source: str = "support_panel"
    # Optional feature attribution when the surface knows it (e.g. "chat",
    # "extraction"). Free-form on purpose — capture points vary.
    feature: Optional[str] = None
    user_id: Optional[str] = None
    team_id: Optional[str] = None
    created_at: datetime.datetime = Field(default_factory=datetime.datetime.utcnow)

    class Settings:
        name = "product_feedback"
        indexes = [
            "sentiment",
            [("created_at", -1)],
        ]
